blog-content/articles/protect-your-future-self-by-encapsulating-test-input-preparation.md

```diff
@@ -0,0 +1,127 @@
+
+# Test Input Initial Encapsulation
+
+Let's extract `Transaction` creation code into `TestTransactions`
+
+:include-file: com/example/portfolio/ProfitCalculatorWithBasicEncapsulationTest.java {title: "ProfitCalculatorTest.java with initial encapsulation"}
+
+:include-java: com/example/portfolio/TestTransactions.java {title: "TestTransactions.createTransaction", entry: "createTransaction"}
+
+Less verbosity. A good start.
+
+Question: Can you still spot the problem?
+
+```spoiler {title: "tap me to see remaining problems"}
+* As we add more properties to `Transaction` class, some of the tests may need to set *extra transaction properties*. 
+* Additionally some of the properties will be completely *irrelevant to the business logic* under a test, yet you will be forced to set them anyway.
+```
+
+# Test Input Encapsulation With Webtau Table Data   
+
+I was dealing with test problems like above for years and eventually came up with a solution that fits my needs.
+I hope it will fit your needs as well. The solution is to use flexible data structure like [Webtau TableData](https://twosigma.github.io/webtau/guide/reference/table-data)
+
+:include-file: com/example/portfolio/ProfitCalculatorWithTableDataTest.java {title: "ProfitCalculatorTest.java with TableData encapsulation", commentsType: "inline"}
+
+Note: [Webtau](https://github.com/twosigma/webtau) open source project started as my answer to common testing problems I encountered. 
+While I am not working at [Two Sigma](https://www.twosigma.com/) anymore I am still contributing and using it on a regular basis.     
+
+:include-java: com/example/portfolio/TestTransactions.java {title: "TestTransactions.createTransactions", commentsType: "inline", entries: ["createTransactions", "genTransactionId"]}
+
+Notice how `createTransactions` defaults values when they are not present?
+As a result of this, tests that don't need say `id` or `lot` are free to ignore them.
+
+:include-java: com/example/portfolio/ProfitCalculatorWithTableDataNoIdNoPriceTest.java {title: "ignored id,lot Transaction properties", commentsType: "inline", entry: "profitShouldBeZeroIfNoLotsSet"}
+
+If later a new required property will be added to `Transaction`, you won't have to change all your existing tests. 
+Instead you will update `createTransaction` with a new default value. 
+The only time you will have to change the tests to use the new property is if the new property is affecting the logic under test.
+
+Essentially your test will use as little data to prove logic works as possible.  
+
+Question: Why it is important to minimize the data in your test?
+
+```spoiler {title: "tap me to see why minimizing data is important"}
+* Test is crucial to understanding *complex business logic*
+* The more data the harder it to comprehend the logic
+* We want business people to look at our data sets (stay tuned to see a good way to expose test data to business)  
```

Comment:
do we want business people to look at our data sets, or do we want them to look at our test cases? 

I really love this, by the way!

Reply:
Data sets. Inputs and output.

Reply:
Hmmm, ok. You are going to go over this in a later article, so that's totally fine by me :)